File: dozor/collectors/logs.py
"""
Container log collector.

Security: Validates all inputs before shell execution.
"""

import re
from datetime import datetime
from typing import Optional

from ..transport import SSHTransport
from ..models import LogEntry
from ..analyzers.log_analyzer import is_bot_scanner_request
from ..validation import validate_service_name, validate_time_duration, sanitize_for_shell


class LogCollector:
    """Collects and parses container logs."""

    # Timestamp patterns for different log formats
    TIMESTAMP_PATTERNS = [
        # ISO format: 2024-01-15T10:30:45.123Z
        (r'^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:?\d{2})?)', '%Y-%m-%dT%H:%M:%S'),
        # Docker compose format: 2024-01-15 10:30:45
        (r'^(\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2})', '%Y-%m-%d %H:%M:%S'),
        # Syslog-like: Jan 15 10:30:45
        (r'^([A-Z][a-z]{2}\s+\d{1,2}\s+\d{2}:\d{2}:\d{2})', '%b %d %H:%M:%S'),
    ]

    # Log level patterns
    LEVEL_PATTERNS = [
        (r'\b(ERROR|FATAL|CRITICAL)\b', 'ERROR'),
        (r'\b(WARN|WARNING)\b', 'WARNING'),
        (r'\b(INFO)\b', 'INFO'),
        (r'\b(DEBUG|TRACE)\b', 'DEBUG'),
        # Postgres specific
        (r'\bERROR:', 'ERROR'),
        (r'\bWARNING:', 'WARNING'),
        (r'\bLOG:', 'INFO'),
        # Node.js/n8n specific
        (r'"level":\s*"error"', 'ERROR'),
        (r'"level":\s*"warn"', 'WARNING'),
        (r'"level":\s*"info"', 'INFO'),
        # Shell/system errors
        (r'Permission denied', 'ERROR'),
        (r'cannot create', 'ERROR'),
        (r'No such file or directory', 'ERROR'),
        (r'command not found', 'ERROR'),
        (r'Segmentation fault', 'ERROR'),
        (r'Killed', 'ERROR'),
        (r'OOM', 'ERROR'),
        (r'out of memory', 'ERROR'),
    ]

    def __init__(self, transport: SSHTransport):
        self.transport = transport

    def get_logs(
        self,
        service: str,
        lines: int = 100,
        since: Optional[str] = None,
    ) -> list[LogEntry]:
        """
        Get logs for a service.

        Args:
            service: Service name (validated)
            lines: Number of lines to fetch
            since: Optional time filter (e.g., "1h", "30m") (validated)

        Returns:
            List of parsed LogEntry objects
        """
        # Validate service name (prevents command injection)
        valid, reason = validate_service_name(service)
        if not valid:
            raise ValueError(f"Invalid service name: {reason}")

        # Validate and sanitize lines count
        if lines < 1 or lines > 10000:
            lines = 100

        cmd_parts = [f'logs --tail {lines} --no-color']

        # Validate since parameter (prevents command injection)
        if since:
            valid, reason = validate_time_duration(since)
            if not valid:
                raise ValueError(f"Invalid time duration: {reason}")
            cmd_parts.append(f'--since {sanitize_for_shell(since)}')

        cmd_parts.append(sanitize_for_shell(service))
        cmd = ' '.join(cmd_parts)

        result = self.transport.docker_compose_command(cmd)

        if not result.success:
            return []

        return self._parse_logs(result.stdout, service)

    def _parse_logs(self, raw_logs: str, service: str) -> list[LogEntry]:
        """Parse raw log output into LogEntry objects."""
        entries = []

        for line in raw_logs.split('\n'):
            if not line.strip():
                continue

            # Skip docker compose prefix if present (e.g., "service-1  | ")
            clean_line = self._strip_compose_prefix(line)

            timestamp = self._extract_timestamp(clean_line)
            level = self._detect_level(clean_line)
            message = self._extract_message(clean_line)

            entries.append(LogEntry(
                timestamp=timestamp,
                level=level,
                message=message,
                service=service,
                raw=line,
            ))

        return entries

    def _strip_compose_prefix(self, line: str) -> str:
        """Remove docker compose service prefix from log line."""
        # Pattern: "service-name-1  | actual log content"
        match = re.match(r'^[\w-]+-\d+\s*\|\s*(.*)$', line)
        if match:
            return match.group(1)
        return line

    def _extract_timestamp(self, line: str) -> Optional[datetime]:
        """Extract timestamp from log line."""
        for pattern, date_format in self.TIMESTAMP_PATTERNS:
            match = re.search(pattern, line, re.IGNORECASE)
            if match:
                ts_str = match.group(1)
                # Clean up timezone for parsing
                ts_str = re.sub(r'Z$', '', ts_str)
                ts_str = re.sub(r'[+-]\d{2}:?\d{2}$', '', ts_str)
                ts_str = re.sub(r'\.\d+$', '', ts_str)

                try:
                    return datetime.strptime(ts_str, date_format)
                except ValueError:
                    continue
        return None

    def _detect_level(self, line: str) -> str:
        """Detect log level from line content."""
        for pattern, level in self.LEVEL_PATTERNS:
            if re.search(pattern, line, re.IGNORECASE):
                return level

        # Default to INFO if no level detected
        return 'INFO'

    def _extract_message(self, line: str) -> str:
        """Extract the message portion of the log line."""
        # Remove timestamp prefix if present
        for pattern, _ in self.TIMESTAMP_PATTERNS:
            line = re.sub(pattern, '', line, count=1).strip()

        # Remove common log level prefixes
        line = re.sub(r'^[\[\(]?(ERROR|WARN|WARNING|INFO|DEBUG|LOG|FATAL)[\]\)]?:?\s*', '', line, flags=re.IGNORECASE)

        return line.strip() or line

    def get_error_logs(
        self,
        service: str,
        lines: int = 100,
        since: Optional[str] = None,
    ) -> list[LogEntry]:
        """Get only error-level logs for a service (excludes bot scanner noise)."""
        all_logs = self.get_logs(service, lines, since)
        return [
            log for log in all_logs
            if log.level in ('ERROR', 'CRITICAL', 'FATAL')
            and not is_bot_scanner_request(log.raw)
        ]

    def search_logs(
        self,
        service: str,
        pattern: str,
        lines: int = 100,
    ) -> list[LogEntry]:
        """Search logs for a specific pattern."""
        all_logs = self.get_logs(service, lines)
        compiled = re.compile(pattern, re.IGNORECASE)
        return [log for log in all_logs if compiled.search(log.raw)]

File: internal/engine/logs.go
package engine

import (
	"context"
	"fmt"
	"regexp"
	"strings"
	"time"
)

// LogCollector gathers and parses container logs.
type LogCollector struct {
	transport *Transport
}

type timestampPattern struct {
	re     *regexp.Regexp
	layout string
}

var timestampPatterns = []timestampPattern{
	{regexp.MustCompile(`^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})`), "2006-01-02T15:04:05"},
	{regexp.MustCompile(`^(\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2})`), "2006-01-02 15:04:05"},
	{regexp.MustCompile(`^([A-Z][a-z]{2}\s+\d{1,2}\s+\d{2}:\d{2}:\d{2})`), "Jan 2 15:04:05"},
}

type levelPattern struct {
	re    *regexp.Regexp
	level string
}

var levelPatterns = []levelPattern{
	{regexp.MustCompile(`\b(ERROR|FATAL|CRITICAL)\b`), "ERROR"},
	{regexp.MustCompile(`\b(WARN|WARNING)\b`), "WARNING"},
	{regexp.MustCompile(`\b(INFO)\b`), "INFO"},
	{regexp.MustCompile(`\b(DEBUG|TRACE)\b`), "DEBUG"},
	{regexp.MustCompile(`\bERROR:`), "ERROR"},
	{regexp.MustCompile(`\bWARNING:`), "WARNING"},
	{regexp.MustCompile(`\bLOG:`), "INFO"},
	{regexp.MustCompile(`"level":\s*"error"`), "ERROR"},
	{regexp.MustCompile(`"level":\s*"warn"`), "WARNING"},
	{regexp.MustCompile(`"level":\s*"info"`), "INFO"},
	{regexp.MustCompile(`Permission denied`), "ERROR"},
	{regexp.MustCompile(`cannot create`), "ERROR"},
	{regexp.MustCompile(`No such file or directory`), "ERROR"},
	{regexp.MustCompile(`command not found`), "ERROR"},
	{regexp.MustCompile(`Segmentation fault`), "ERROR"},
	{regexp.MustCompile(`Killed`), "ERROR"},
	{regexp.MustCompile(`OOM`), "ERROR"},
	{regexp.MustCompile(`out of memory`), "ERROR"},
}

// logsSinceWindow limits log collection to recent entries only,
// preventing old log lines from being reported as current issues.
const logsSinceWindow = "1h"

// GetLogs fetches logs for a service.
func (c *LogCollector) GetLogs(ctx context.Context, service string, lines int, errorsOnly bool) []LogEntry {
	tailN := lines
	if tailN <= 0 {
		tailN = 100
	}

	cmd := fmt.Sprintf("logs --tail %d --since %s --timestamps %s 2>&1", tailN, logsSinceWindow, service)
	res := c.transport.DockerComposeCommand(ctx, cmd)
	if !res.Success {
		return nil
	}

	entries := parseLogLines(res.Stdout, service)

	if errorsOnly {
		filtered := make([]LogEntry, 0)
		for _, e := range entries {
			if e.IsErrorLevel() {
				filtered = append(filtered, e)
			}
		}
		return filtered
	}

	// Filter bot scanner noise
	entries = filterBotScanner(entries)

	return entries
}

// GetErrorLogs fetches only error logs.
func (c *LogCollector) GetErrorLogs(ctx context.Context, service string, lines int) []LogEntry {
	return c.GetLogs(ctx, service, lines, true)
}

func parseLogLines(output, service string) []LogEntry {
	lines := strings.Split(output, "\n")
	entries := make([]LogEntry, 0, len(lines))

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		// Strip docker compose prefix "service-name  | "
		msg := line
		if idx := strings.Index(line, " | "); idx > 0 {
			msg = line[idx+3:]
		}

		entry := LogEntry{
			Service: service,
			Raw:     line,
			Message: msg,
			Level:   detectLevel(msg),
		}
		entry.Timestamp = parseTimestamp(msg)
		entries = append(entries, entry)
	}

	return entries
}

func detectLevel(line string) string {
	for _, p := range levelPatterns {
		if p.re.MatchString(line) {
			return p.level
		}
	}
	return "INFO"
}

func parseTimestamp(line string) *time.Time {
	for _, p := range timestampPatterns {
		if m := p.re.FindString(line); m != "" {
			if t, err := time.Parse(p.layout, m); err == nil {
				return &t
			}
		}
	}
	return nil
}

// Bot scanner paths to filter out noise
var botScannerPaths = []*regexp.Regexp{
	regexp.MustCompile(`/SDK/webLanguage`),
	regexp.MustCompile(`/sdk/`),
	regexp.MustCompile(`/wp-admin`),
	regexp.MustCompile(`/wp-login`),
	regexp.MustCompile(`/wp-content`),
	regexp.MustCompile(`/wp-includes`),
	regexp.MustCompile(`/xmlrpc\.php`),
	regexp.MustCompile(`\.php$`),
	regexp.MustCompile(`/phpmyadmin`),
	regexp.MustCompile(`/phpMyAdmin`),
	regexp.MustCompile(`/pma`),
	regexp.MustCompile(`/\.env`),
	regexp.MustCompile(`/\.git`),
	regexp.MustCompile(`/config\.json`),
	regexp.MustCompile(`/\.aws`),
	regexp.MustCompile(`/\.ssh`),
	regexp.MustCompile(`/admin`),
	regexp.MustCompile(`/administrator`),
	regexp.MustCompile(`/manager`),
	regexp.MustCompile(`/console`),
	regexp.MustCompile(`\.bak$`),
	regexp.MustCompile(`\.backup$`),
	regexp.MustCompile(`\.sql$`),
	regexp.MustCompile(`\.tar\.gz$`),
	regexp.MustCompile(`\.zip$`),
	regexp.MustCompile(`/api/v\d+/swagger`),
	regexp.MustCompile(`/swagger`),
	regexp.MustCompile(`/actuator`),
	regexp.MustCompile(`/metrics`),
	regexp.MustCompile(`/debug`),
	regexp.MustCompile(`/cgi-bin`),
	regexp.MustCompile(`/shell`),
	regexp.MustCompile(`/cmd`),
	regexp.MustCompile(`/eval`),
	regexp.MustCompile(`/exec`),
}

func filterBotScanner(entries []LogEntry) []LogEntry {
	out := make([]LogEntry, 0, len(entries))
	for _, e := range entries {
		isBot := false
		for _, p := range botScannerPaths {
			if p.MatchString(e.Message) {
				isBot = true
				break
			}
		}
		if !isBot {
			out = append(out, e)
		}
	}
	return out
}
